runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/GrpcGetDataStream.java

```diff
@@ -192,11 +205,35 @@ public void refreshActiveWork(Map<String, List<KeyedGetDataRequest>> active) {
                 .addRequests(request));
       }
     }
+    for (Map.Entry<String, List<HeartbeatRequest>> entry : heartbeats.entrySet()) {
+      for (HeartbeatRequest request : entry.getValue()) {
+        // Calculate the bytes with some overhead for proto encoding.
+        long bytes = (long) entry.getKey().length() + request.getSerializedSize() + 10;
+        if (builderBytes > 0
+            && (builderBytes + bytes > AbstractWindmillStream.RPC_STREAM_CHUNK_SIZE
+                || builder.getRequestIdCount() >= streamingRpcBatchLimit)) {
+          send(builder.build());
+          builderBytes = 0;
+          builder.clear();
+        }
+        builderBytes += bytes;
+        builder.addComputationHeartbeatRequest(
+            ComputationHeartbeatRequest.newBuilder()
+                .setComputationId(entry.getKey())
```

Comment:
this adds a new computationheartbeatrequest for each heartbeat, we should just have one per computation with many heartbeats inside them (add test to catch that).

Reply:
GrpcWindmillServerTest.testStreamingGetDataHeartbeatsAsHeartbeatRequests should now do the right thing. Before, the test was exploding any requests sent before verifying, so we always got one key per ComputationHeartbeatRequest.